api/auth.proto

```diff
@@ -0,0 +1,53 @@
+syntax = "proto3";
+
+package envoy.api.v2;
+
+message AuthAction {
+
+  enum ActionType {
+    // Request matches all rules are allowed, otherwise denied.
+    ALLOW = 0;
+    // Request matches all rules or missing required auth fields are denied,
+    // otherwise allowed.
+    DENY = 1;
+  }
+
+  ActionType action_type = 1;
+
+  // Check peer identity using X.509 certificate.
+  message X509Assert {
+    // SHA256 hash of peer certificate.
+    string sha256_hash = 1;
+    // Subject of peer certificate.
+    string subject = 2;
+    // Subject alt name of peer certificate.
+    string subject_alt_name = 3;
+  }
+
+  enum RuleType {
+    X509 = 0;
+  }
+
+  // Rule contains list of asserts of the same type.
+  message Rule {
+    RuleType rule_type = 1;
```

Comment:
Do we need the enum? Isn't it implicit from the oneof below?

Reply:
i think all list item has the same type in the list, and ensure the type is correct.
if there is no type info, we can't know which type to use.

or do you mean the list can contain different type ?